### PR TITLE
Introduce janitor for IBM Cloud

### PR DESCRIPTION
**Pull request by Amulyam24**
This PR adds a new janitor for IBM Cloud resources.

### CODE REVIEW

**Comment by Amulyam24** [comment]
cc @mkumatag @Rajalakshmi-Girish

File: go.mod
Comment (mkumatag) [inline go.mod:8]: seems like the prow job is still pointing to older version of golang and k8s, lets rollback version to older version of k8s bits.

File: ibmcloud-janitor/OWNERS
Comment (mkumatag) [inline ibmcloud-janitor/OWNERS:9]: add newline

File: cmd/ibmcloud-janitor-boskos/OWNERS
Comment (mkumatag) [inline cmd/ibmcloud-janitor-boskos/OWNERS:9]: add newline

File: ibmcloud-janitor/resources/serviceid.go
Comment (Amulyam24) [inline ibmcloud-janitor/resources/serviceid.go:64]: @mkumatag, the failure scenarios can be handled as mentioned below. Please share your thoughts in case I missed anything here.
1. Accidental deletion of API key assigned to the resource when the state of the resource is

_Free/Busy_: 
- The triggered test would fails while authenticating with IBM Cloud since the API key no longer exists
-  The state of the resource is set to dirty once the test ends after which the janitor picks it up, cleans it and updates the resource with a new API Key followed by setting its state to free 
- Subsequent test runs will eventually pass by using the new API key for authentication

_Any other state_: 
- Reaper sets the state of the resource to dirty irrespective of the current state after a certain time period 
- Janitor cleans up the resource and updates the API key with a new value and sets it to state free after which it can be used by any test.


2. Janitor fails to create a new API key or updating the resource with the new key
- On failure, the resource would still be in a dirty state(can not be used by any project/owner) and janitor will be responsible for creating a new API key in subsequent runs before finally setting its state to free

Comment (mkumatag) [reply ibmcloud-janitor/resources/serviceid.go:64]: this sgtm

Comment (Rajalakshmi-Girish) [reply ibmcloud-janitor/resources/serviceid.go:64]: @Amulyam24 In the second failure scenario, will the Janitor not be infinitely trying to Create a new API key? (Hypothetically, if the reason for APIKey creation failure is not resolved)
Hope Janitor or the Test has a timeout value.

Comment (Amulyam24) [reply ibmcloud-janitor/resources/serviceid.go:64]: @Rajalakshmi-Girish, if there has been an unforeseen failure while trying to create a new API key, the resource would continuously be transitioning from cleaning <--> dirty state (set by reaper after 30m expiration time) thereby making it unusable by any test. I think resolving it would require probing the logs to find the actual cause and fixing it.

Comment (Rajalakshmi-Girish) [reply ibmcloud-janitor/resources/serviceid.go:64]: @Amulyam24 Yes you are right, the second failure scenario happens with a resource in `dirty` state.

But if the cause for `create a new API key` failure is not in our scope of resolution, ( For example, if it is something to do with PowerVS) will Janitor infinitely keep trying to `create a new API Key`? or is there a timeout that Janitor has for dirty resources that it handles?

Comment (Amulyam24) [reply ibmcloud-janitor/resources/serviceid.go:64]: > will Janitor infinitely keep trying to create a new API Key? 

Yes, until it is able to create one and set the resource to free state. 

> or is there a timeout that Janitor has for dirty resources that it handles?

There is no timeout set for Janitor if it continuously fails to clean up a dirty resource. It will keep looking for dirty resources until they are set to free.

**Comment by Amulyam24** [comment]
Hi @alvaroaleman @stevekuznetsov @fejta, a ping for review, PTAL :)

File: ibmcloud-janitor/resources/serviceid.go
Comment (Rajalakshmi-Girish) [reply ibmcloud-janitor/resources/serviceid.go:71]: > Yes, until it is able to create one and set the resource to free state.

If Janitor is unable to create APIKey for quite some time, will it report the failure?

Should we consider the janitor to have a timeout for handling each dirty resource? Else, scenarios like the second one might lead to a situation where resource remains in `dirty` state and might not be reported.

Also, there might be a situation where all resources posted to `Boskos` might move to `dirty` state and eventually lead to test failure (as it would be unable to find a free resource to run the test).

File: cmd/ibmcloud-janitor-boskos/OWNERS
Comment (chaodaiG) [inline cmd/ibmcloud-janitor-boskos/OWNERS:4]: 👍

File: ibmcloud-janitor/OWNERS
Comment (chaodaiG) [inline ibmcloud-janitor/OWNERS:1]: not a comment on this file, I would expect these to be under `pkg` directory from top level

**Review by chaodaiG — COMMENTED** [review]
Thank you for adding OWNERS file, I can approve once you got lgtm from someone who has enough context

**Comment by mkumatag** [comment]
> Thank you for adding OWNERS file, I can approve once you got lgtm from someone who has enough context

I will do that honor :)

File: ibmcloud-janitor/resources/serviceid.go
Comment (mkumatag) [inline ibmcloud-janitor/resources/serviceid.go:41]: wondering if we can use some k8s utilities to generate the random string here.

File: ibmcloud-janitor/resources/serviceid.go
Comment (mkumatag) [inline ibmcloud-janitor/resources/serviceid.go:64]: if we are already searching the entries by name then why do we need this comparison again here?

File: cmd/ibmcloud-janitor-boskos/main.go
Comment (mkumatag) [inline cmd/ibmcloud-janitor-boskos/main.go:25]: can we add a resource field with all the messages we are logging for the better logging.

e.g:

```goalng
resourceLogger := log.WithFields(log.Fields{"resource": resourceName})
resourceLogger.WithField("Service ID: ", serviceID.Name).Info("found the target service ID")
```

File: ibmcloud-janitor/account/account.go
Comment (mkumatag) [inline ibmcloud-janitor/account/account.go:27]: Wondering if we can add an environment variable here to override this value runtime

File: cmd/ibmcloud-janitor-boskos/main.go
Comment (mkumatag) [inline cmd/ibmcloud-janitor-boskos/main.go:53]: is this `boskos.Acquire` randomly pick or goes in order?

if goes in order what happens in the next loop if it fails to cleanup the resource

File: ibmcloud-janitor/resources/clean.go
Comment (mkumatag) [inline ibmcloud-janitor/resources/clean.go:48]: I see its NoOps if user specify list of resources in the command line e.g: `--resource-type=powervs-service-instance,vpc-service,classic`, nothing happens in this function with new info/warning message.

It will be good to add some warning message if user specify something else or allow user to use only the `powervs-service-instance,vpc-service`

File: ibmcloud-janitor/resources/types.go
Comment (mkumatag) [inline ibmcloud-janitor/resources/types.go:25]: Is this a interface generic or only for the powervs resources?

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: ibmcloud-janitor/resources/types.go
Comment (mkumatag) [inline ibmcloud-janitor/resources/types.go:65]: I'm wondering instead of locking to a 1-1 mapping, can we compare with the prefix, e.g: all the resources type with `powervs-*` can be considered as powervs type and this way we can allow boskos resources to be clubbed in different sets for the better management if required

File: ibmcloud-janitor/README.md
Comment (Amulyam24) [reply ibmcloud-janitor/README.md]: Changed.

File: ibmcloud-janitor/README.md
Comment (Amulyam24) [reply ibmcloud-janitor/README.md]: Changed to `k8s-boskos-powervs-<loc>-00`.

File: ibmcloud-janitor/README.md
Comment (Amulyam24) [reply ibmcloud-janitor/README.md:1]: Enhanced the Readme, PTAL.

File: go.mod
Comment (Amulyam24) [reply go.mod:17]: This was updated by an earlier change in client-go version.

File: ibmcloud-janitor/resources/serviceid.go
Comment (Amulyam24) [reply ibmcloud-janitor/resources/serviceid.go]: Changed to use k8s utility.

File: ibmcloud-janitor/resources/serviceid.go
Comment (Amulyam24) [reply ibmcloud-janitor/resources/serviceid.go]: yes, we don't need this. Changed it.

File: ibmcloud-janitor/resources/clean.go
Comment (Amulyam24) [reply ibmcloud-janitor/resources/clean.go:41]: Changed and added a warning.

File: ibmcloud-janitor/README.md
Comment (Amulyam24) [reply ibmcloud-janitor/README.md]: Added some details for using API key and doc references for the Powervs resources.

File: cmd/ibmcloud-janitor-boskos/main.go
Comment (Amulyam24) [reply cmd/ibmcloud-janitor-boskos/main.go:53]: Acquire follows FIFO approach. The unsuccessfully cleaned up dirty resource will be eventually picked up by janitor in the next loop for cleanup.

Comment (mkumatag) [reply cmd/ibmcloud-janitor-boskos/main.go:53]: but happens for the next loop, is that resource again sit first in the queue or added in the end?

File: ibmcloud-janitor/resources/types.go
Comment (Amulyam24) [reply ibmcloud-janitor/resources/types.go]: yes, this could be better. Changed it.

File: cmd/ibmcloud-janitor-boskos/main.go
Comment (Amulyam24) [reply cmd/ibmcloud-janitor-boskos/main.go:53]: It might get added to the end of the queue since the resource would transition from `dirty --> cleaning --> dirty`.

File: ibmcloud-janitor/OWNERS
Comment (Amulyam24) [reply ibmcloud-janitor/OWNERS:1]: > not a comment on this file, I would expect these to be under `pkg` directory from top level

Hi @chaodaiG, I'm not sure what you meant here? I don't see a `pkg` directory in the top level !?

**Review by mkumatag — APPROVED** [review]
/lgtm

/assign @chaodaiG

**Comment by chaodaiG** [comment]
@Amulyam24 please take a look at my comment at https://github.com/kubernetes-sigs/boskos/pull/122#discussion_r871478607

**Comment by Amulyam24** [comment]
@chaodaiG @mkumatag , Addressed all review comments, PTAL.

**Review by chaodaiG — APPROVED** [review]

**Pull request merged**
